### ISSUE TITLE
Refactor everything

### ISSUE THREAD

**Issue by okitec** [dieechtenilente/alcoholtestapp#42]
Name       | Definition
-------------|------------------------------------
Recipe      | Array of `Ingredient`s
Recipe.mass() | Sum of the alcohol masses of the ingredients
Ingredient | `(name, alcohol content, quantity [ml])`
Ingredient.mass() | Returns the mass A [gram] of the alcohol as defined by `A = quantity [ml] * alcohol content * density of alcohol (0.8 g/ml)`.
Drink        | A consumed `Recipe` with the constructor `new Drink(Recipe r);`
Drink        | `(name, recipe, consumption time, remaining factor)`
Drink.mass() | Returns the remaining unprocessed mass of the drink's alcohol after exponential decline: `remaining mass = remaining factor * recipe.mass()`.
Drink.dec() | Decreases the remaining factor by the average decline rate 0.00015 and returns the new remaining factor. **This is the only call that modifies anything. Call it once every hour for the first drink in every User's list. If the result is <= 0, remove the drink; also, if it is < 0, call dec(abs(result)) for the next drink in the list, and so on. This way, all of the processing capacity of the liver is accounted for. Else, we'd remove only one drink where we should remove more.**
User         |  `(name, sex, age, height, mass, drinks)`
User.waterQuantity() | (*private*) Returns the total volume of the water in the body (GKW on Wikipedia (BAK/Watson-Formel)).
User.reductionFactor() | Returns the percentage of the body reached by the alcohol as defined by `r = blood density (1.055g/ml) * waterQuantity() / 0.8 * mass`.
User.alcoholMass() | Returns the mass of the remaining alcohol in the blood defined as the sum of `drink[i].mass()`.
User.bac() | Returns the BAC of the user as defined by Widmark: `bac =  alcoholMass() / mass * reductionFactor()`.

**Comment by dieechtenilente** [comment]
`Recipe`, `Ingredient` added with 4e87491c1f57efffe749209cc36b9016844530da

**Comment by dieechtenilente** [comment]
-> #40, #41

**Comment by okitec** [comment]
Maximum nitpick mode engaged, I hope you don't mind me criticising every little bit. I only care for code quality. Make it the best app (die **BWESTE**)!

**Comment by dieechtenilente** [comment]
It's okay. THIS WILL BE THE BEST APP!!!!11 SCHAUNS HER!

**Comment by okitec** [comment]
I'm concerned the remaining factor/dec logic might be wrong. Each tick, we want to decrement the first drink's alcohol mass by 0.00015 **times the total alcohol mass**, not times the first alcohol mass only. But we use the remaining factor here, so I don't know if it is what we want. Maths!

**Comment by dieechtenilente** [comment]
I think, my original logic is correct. Say you drank a drink (lol) at time A and it has alcohol mass M in it, so you use `alcohol mass` * 3600000 to deplete ~0,1 ‰ per hour [[like in line 236](https://github.com/dieechtenilente/alcoholtestapp/blob/f0e6b4ec5c81b03b31985c6e7a173be474963bea/app/src/main/java/de/klaushackner/breathalyzer/MainActivity.java): `expireTime = takingTime + Math.round(bac * 36000000)].` where `bac` is from this methode:

```java
public static double getBac(Mixture mixture, User user) {
  if (user.isMale()) {
    double R = 2.447 - 0.09516 * user.getAge() + 0.1074 * user.getHeight() + 0.3362 * user.getWeight();
  } else {
    double R = -2.097 + 0.1069 * user.getHeight() + 0.2466 * user.getWeight();
  }

  r = (1.055 * R) / (0.8 * user.getWeight());
  return (mixture.getAmount() * mixture.getPercentage() * 0.8) / (user.getWeight() * r);
}
```

So you need only a factor so you can get the remaining alcohol mass from this drink at a given point in time.

**Comment by okitec** [comment]
Why would the mass decrease when you multiply it with a huge value?

My logic would be fine if we stored a remaining alcohol mass in Drink, instead of a remainingFactor. I suggest we use the new system and revert if it turns out to be wrong.

**Comment by dieechtenilente** [comment]
Why store something in the drink anyway? We have a taking time and the alc-mass. Everything could be calculated on the fly.

**Comment by okitec** [comment]
You drink a drink with 50mg alcohol at time 0h.
You drink a drink with 20mg alcohol at time 1h.
How much grams of alcohol are there at time 2h?

Well, you are probably right, since there's the distributive law, I derped the night I wrote the proposal.

```
reduced alc mass at time t [hours] = t * 0.00015 * total alc mass
    = t * 0.00015 * alc mass 1 + t * 0.00015 * alc mass 2 + ...
```

For this to work, we need to set a time epoch t = 0. We mustn't consider any taking times except the one of the first drink. That's probably why I thought this wouldn't work.
